crates/oxc_linter/src/rules/nextjs/no_duplicate_head.rs

```diff
@@ -0,0 +1,215 @@
+    AstKind,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::{GetSpan, Span};
+
+use crate::{context::LintContext, rule::Rule};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint-plugin-next(no-duplicate-head):")]
+#[diagnostic(severity(warning), help("Do not include multiple instances of `<Head/>`. See: https://nextjs.org/docs/messages/no-duplicate-head"))]
+struct NoDuplicateHeadDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoDuplicateHead;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Prevent duplicate usage of <Head> in pages/_document.js.
+    ///
+    /// ### Why is this bad?
+    /// This can cause unexpected behavior in your application.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// import Document, { Html, Head, Main, NextScript } from 'next/document'
+    /// class MyDocument extends Document {
+    ///   static async getInitialProps(ctx) {
+    ///   }
+    ///   render() {
+    ///     return (
+    ///       <Html>
+    ///         <Head />
+    ///         <body>
+    ///           <Main />
+    ///           <NextScript />
+    ///         </body>
+    ///       </Html>
+    ///    )
+    ///  }
+    ///}
+    ///export default MyDocument
+    /// ```
+    NoDuplicateHead,
+    correctness
+);
+
+impl Rule for NoDuplicateHead {
+    fn run_once(&self, ctx: &LintContext) {
+        let mut document_import_name = None;
+        let nodes = ctx.semantic().nodes();
+        for node in nodes.iter() {
+            match node.kind() {
+                oxc_ast::AstKind::ImportDeclaration(decl) => {
+                    if decl.source.value == "next/document" {
+                        let document_import = decl.specifiers.as_ref().and_then(|specifiers| {
+                            specifiers.iter().find(|spec| matches!(spec, ImportDefaultSpecifier(_)))
+                        });
+                        if let Some(ImportDefaultSpecifier(document_import_specifier)) =
+                            document_import
+                        {
+                            document_import_name =
+                                Some(document_import_specifier.local.name.clone());
+                        }
+                    }
+                }
+                oxc_ast::AstKind::ReturnStatement(stmt) => {
+                    let document_class_id = nodes.ancestors(node.id()).find(|node_id| {
+                        matches!(nodes.get_node(*node_id).kind(),
+                        AstKind::Class(class)
+                            if class
+                                .super_class
+                                .as_ref()
+                                .and_then(|sc| sc.as_identifier())
+                                .map(|id| &id.name)
+                                == document_import_name.as_ref())
+                    });
```

Comment:
I think we can have a simpler way to find the class, for example by `ctx.semantic.classes`

Reply:
Not sure if this would be faster, you still need to figure out if the class is the ancestor of current `ReturnStatement`